src/lib.rs

```diff
@@ -45,869 +57,454 @@ extern crate alloc;
-        v.write_u5(u5::try_from(*b & 0x1f).expect("can't be out of range, max. 31"))?;
-    }
-    Ok(())
-}
-
-fn polymod(values: &[u5]) -> u32 {
-    let mut chk: u32 = 1;
-    let mut b: u8;
-    for v in values {
-        b = (chk >> 25) as u8;
-        chk = (chk & 0x01ff_ffff) << 5 ^ (u32::from(*v.as_ref()));
+pub fn encode_force_bech32m(
+    network: Network,
+    witness_version: WitnessVersion,
+    data: &[u8],
+) -> String {
+    use primitives::iter::{ByteIterExt, Fe32IterExt};
+
+    data.iter()
+        .copied() // iterate over bytes
+        .bytes_to_fes() // convert bytes to field elements in-line
+        .with_witness_version(witness_version)
+        .checksum::<Bech32m>()
+        .with_checksummed_hrp_from_network(network)
+        .hrp_char_from_network(network)
+        .collect()
+}
+
+/// The "null checksum" used on bech32 strings for which we want to do no checksum checking
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum NoChecksum {}
+
+/// The bech32 checksum algorithm, defined in BIP-173.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Bech32 {}
+
+/// The bech32m checksum algorithm, defined in BIP-350.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Bech32m {}
+
+/// The codex32 checksum algorithm, defined in BIP-93.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum Codex32 {}
+
+impl Checksum for NoChecksum {
+    type MidstateRepr = PackedNull;
+    const CHECKSUM_LENGTH: usize = 0;
+    const GENERATOR_SH: [PackedNull; 5] = [PackedNull; 5];
+    const TARGET_RESIDUE: PackedNull = PackedNull;
+}
+
+// Bech32[m] generator coefficients, copied from Bitcoin Core src/bech32.cpp
+const GEN: [u32; 5] = [0x3b6a_57b2, 0x2650_8e6d, 0x1ea1_19fa, 0x3d42_33dd, 0x2a14_62b3];
 
-        for (i, item) in GEN.iter().enumerate() {
-            if (b >> i) & 1 == 1 {
-                chk ^= item;
-            }
-        }
+impl Checksum for Bech32 {
+    type MidstateRepr = u32;
+    const CHECKSUM_LENGTH: usize = 6;
+    const GENERATOR_SH: [u32; 5] = GEN;
+    const TARGET_RESIDUE: u32 = 1;
+}
+// Same as Bech32 except TARGET_RESIDUE is different
+impl Checksum for Bech32m {
+    type MidstateRepr = u32;
+    const CHECKSUM_LENGTH: usize = 6;
+    const GENERATOR_SH: [u32; 5] = GEN;
+    const TARGET_RESIDUE: u32 = 0x2bc830a3;
+}
+
+impl Checksum for Codex32 {
+    type MidstateRepr = u128;
+    const CHECKSUM_LENGTH: usize = 13;
+    // Copied from BIP-93
+    const GENERATOR_SH: [u128; 5] = [
+        0x19dc500ce73fde210,
+        0x1bfae00def77fe529,
+        0x1fbd920fffe7bee52,
+        0x1739640bdeee3fdad,
+        0x07729a039cfc75f5a,
+    ];
+    const TARGET_RESIDUE: u128 = 0x10ce0795c2fd1e62a;
+}
+
+/// True if program length is either 20 bytes (p2wpkh) or 32 bytes (p2wsh).
+/// ref: [BIP-141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#user-content-Witness_program)
+#[cfg(feature = "alloc")]
```

Comment:
what part of this function requires an allocator?

Reply:
This is to prevent clippy warning when "alloc" is not enabled, I added a comment to it to say as such.